Add calculation for faerie fire feral

waiwai2233 opened the pull request:
This is my first time contributing to such a large project and I understand that I may have made some mistakes.

This pull request contains the following changes related to fearie fire (bear form):

Added "Curse of Elements" effect in damage debuffs
Added "Misery" effect in stat debuffs
Added spell hit calculation in main.py, taking into account debuffs like misery. Player's hit is calculated using the formula `buffed_spell_hit=unbuffed_spell_hit + 1 * ('heroic_presence' in raid_buffs) + 40 / 26.23 * ('hit_food' in consumables)` Encounter hit is calculated using the formula `encounter_spell_hit = (buffed_spell_hit + 3 * ('misery' in stat_debuffs))`
Added spell multiplier, taking into account Heart of the Crusader / Master Poisoner
Modified the faerie_fire() function in the Player class to account for the damage done by Faerie Fire in bear form using formula `faerie_dmg = (0.15 * bear_ap + 1.) * (1 + 0.13 * curse_of_elements) * spell_damage_multiplier`

Please review my changes and let me know if there are any issues or concerns and I look forward to contributing more to this project in the future.

waiwai2233 committed:
Add calculation for faerie fire feral

Add curse of elements in damage debuffs.
Add misery in stat debuffs.
Add spell hit calculation in main.py.
Modifed faerie_fire function in player class to account for damage done of faerie fire in bear form.

NerdEgghead commented on main.py:
```diff
@@ -271,8 +273,9 @@
                  'value': 'jotc'
              },
              {'label': 'Judgment of Wisdom', 'value': 'jow'},
+             {'label': 'Misery / Improved Fearie Fire', 'value': 'misery'},
```
Small typo, should be "Faerie" not "Fearie"

NerdEgghead commented on player.py:
```diff
@@ -51,6 +51,8 @@ def __init__(
             ap_mod (float): Total multiplier for Attack Power in Cat Form.
             agility (int): Fully raid buffed Agility attribute.
             hit_chance (float): Chance to hit as a fraction.
+            spell_hit_chance (float): Chance for spells (farie fire feral) 
```
small typo again, should be "Faerie Fire" not "farie fire"

NerdEgghead commented on player.py:
```diff
@@ -1022,9 +1033,18 @@ def faerie_fire(self):
+
+        self.dmg_breakdown['Faerie Fire (Bear)']['casts'] += 1
+        miss = (np.random.rand() < self.spell_miss_chance)
+        damage_on_hit = self.faerie_fire_hit
```
Faerie Fire (Bear) can crit, so this calculation is insufficient. You'll need to store the player's spell crit as well and then roll for crits using the calc_yellow_damage() helper function.

NerdEgghead commented on player.py:
```diff
@@ -1022,9 +1033,18 @@ def faerie_fire(self):
+        self.dmg_breakdown['Faerie Fire (Bear)']['casts'] += 1
+        miss = (np.random.rand() < self.spell_miss_chance)
+        damage_on_hit = self.faerie_fire_hit
+        self.gen_log('Faerie Fire (Bear)', damage_on_hit, miss, False, False)
```
Should wrap this in an `if self.log:` block for consistency with the rest of the code.

NerdEgghead commented on player.py:
```diff
@@ -202,6 +206,8 @@ def calc_miss_chance(self):
         )
         self.dodge_chance = 0.01 * (6.5 - dodge_reduction)
+        spell_miss_reduction = min(self.spell_hit_chance * 100, 17.0)
+        self.spell_miss_chance = 0.01 * (17.0 - spell_miss_reduction)
```
What's missing here is that dynamic changes to Hit Rating during a fight (from trinket procs etc.) or more importantly, Hit Rating adjustments within `calc_stat_weights`, won't correctly propagate to the spell miss chance.

NerdEgghead reviewed (changes requested):
Thanks for the contribution! Looks good as a starting point, but the two important missing pieces are (a) propagating dynamic Hit Rating changes to the stored spell miss chance, and (b) implementing spell crits on Faerie Fire.

waiwai2233 committed:
Corrected spelling error

waiwai2233 committed:
Add calculation for spell crit

waiwai2233 committed:
Add assumption for spell crit suppression

Santiclause commented on main.py:
```diff
@@ -258,8 +258,10 @@
                  'label': 'Blood Frenzy / Savage Combat',
                  'value': 'blood_frenzy'
              },
+             {'label': "Curse of Elements / Earth and Moon / Ebon Plaguebringer",
+              'value': 'curse_of_elements'},
```
for a multiline dict like this, you should format it like the dict above
```suggestion
             {
                 'label': "Curse of Elements / Earth and Moon / Ebon Plaguebringer",
                 'value': 'curse_of_elements'
             },
```

Santiclause commented on main.py:
```diff
@@ -271,8 +273,12 @@
              },
              {'label': 'Judgment of Wisdom', 'value': 'jow'},
+             {'label': 'Misery / Improved Faerie Fire', 'value': 'misery'},
+             {'label': 'Shadow Mastery / Improved Scorch / Winter\'s Chill',
+              'value': 'shadow_mastery'
```
same suggestion
```suggestion
             {
                 'label': 'Shadow Mastery / Improved Scorch / Winter\'s Chill',
                 'value': 'shadow_mastery'
```

waiwai2233 replied:
okok I will change it right now, btw, do you know where i can find info for spell resistance?

Santiclause reviewed (commented):
some pitnicky suggestions - also, I don't see 5% spell crit aura from boomie/ele sham listed anywhere. not sure if that oversight is intentional or not

waiwai2233 committed:
Change format for multi-line dict

waiwai2233 committed:
Add moonkin aura in raid buffs

waiwai2233 committed:
Modified sim_utils.py and player.py

Add a new function calc_spell_damage to sim_utils.py to calculate
spell damage adjusted for level based resistance.
Change the faerie_fire implementation in player.py to call
calc_spell_damage.

waiwai2233 committed:
Change explanation of function in sim_utils.py

waiwai2233 committed:
Add enrage condition
